File: chapter4/codemaker88.java
package Chapter4;

import java.util.function.Function;
import java.util.stream.Stream;

/**
 * Created by codemaker88 on 2016-11-04.
 */
public class Chapter4 {
    public static void main(String[] args)
    {
        Bibimbap bibimbap = new Bibimbap();
        bibimbap.addMaterial(재료추가::콩나물, 재료추가::고추장);
        System.out.println(bibimbap.mix());
        bibimbap.addMaterial(재료추가::참치, 재료추가::마요네즈, 재료추가::스팸);
        System.out.println(bibimbap.mix());
    }

    static class Bibimbap {
        private static final String NAME = "비빔밥";
        private Function<String, String> material;

        public String mix() {
            return material.apply(NAME);
        }

        public void addMaterial(final Function<String, String>... materials) {
            material = Stream.of(materials)
                    .reduce(Function::compose)
                    .orElse(material -> material);
        }
    }

    static class 재료추가 {

        static String 고추장(String materials) {
            return add("고추장", materials);
        }

        static String 콩나물(String materials) {
            return add("콩나물", materials);
        }

        static String 마요네즈(String materials) {
            return add("마요네즈", materials);
        }

        static String 참치(String materials) {
            return add("참치", materials);
        }

        static String 스팸(String materials) {
            return add("스팸", materials);
        }

        static String add(String material1, String material2) {
            return material1 + " " + material2;
        }
    }
}

File: chapter4/codemaker88.scala
import java.io.File
import java.nio.file.Files
import java.nio.file.Paths
import java.nio.file.Path
import scala.collection.JavaConverters
import scala.collection.Iterator

class Chapter4 {
  def solve() {
    var bibimbap = new Bibimbap()
    bibimbap.addMaterial(콩나물.추가, 고추장.추가)
    println(bibimbap.mix())
    
    bibimbap.addMaterial(스팸.추가, 마요네즈.추가, 소고기.추가, 고추장.추가)
    println(bibimbap.mix())
  }
  
  class Bibimbap {
    var 재료들 : Function1[String, String] = null
    
    def addMaterial(materialList : Function1[String, String]*) {
      재료들 = materialList.reduceRight(_ compose _)
    }
    
    def mix() : String = {
      return 재료들.apply("비빔밥");
    }
    
  }
  
  sealed abstract case class 재료(재료명 : String) {
    
    def 추가(재료들 : String) : String = {
      return this.재료명 + " " + 재료들
    }
  }
  
  object 고추장 extends 재료("고추장")
  object 소고기 extends 재료("소고기")
  object 콩나물 extends 재료("콩나물")
  object 스팸 extends 재료("스팸")
  object 참치 extends 재료("참치")
  object 마요네즈 extends 재료("마요네즈")
}